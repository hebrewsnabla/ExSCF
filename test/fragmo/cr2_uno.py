from pyscf import gto, scf
from fch2py import fch2py
from ortho import check_orthonormal
from pyscf import lib
import numpy as np
import os
from py2fch import py2fch
from uno import uno
from construct_vir import construct_vir

lib.num_threads(1)

mol = gto.M()
#   2 atom(s)
mol.atom = '''
Cr1             0.00000000        0.00000000        0.00000000
Cr2             0.00000000        0.00000000        2.00000000
'''

mol.basis = {
'Cr1': gto.basis.parse('''
Cr    S
    0.254477807E+06    0.242904776E-03
    0.381317971E+05    0.188435236E-02
    0.867529306E+04    0.980095844E-02
    0.245500998E+04    0.398250087E-01
    0.799162178E+03    0.129405437E+00
    0.286900215E+03    0.306290018E+00
    0.111254132E+03    0.434628350E+00
    0.438641526E+02    0.224695629E+00
Cr    S
    0.279326692E+03   -0.243633579E-01
    0.862747324E+02   -0.115114954E+00
    0.135557561E+02    0.550922663E+00
    0.569781128E+01    0.536113547E+00
Cr    S
    0.856365826E+01   -0.371230177E+00
    0.139882968E+01    0.116811695E+01
Cr    S
    0.572881711E+00    0.100000000E+01
Cr    S
    0.900961713E-01    0.100000000E+01
Cr    S
    0.341258851E-01    0.100000000E+01
Cr    P
    0.130643989E+04    0.282928007E-02
    0.309253114E+03    0.227766292E-01
    0.989962740E+02    0.105645619E+00
    0.367569165E+02    0.299499449E+00
    0.145666571E+02    0.477062454E+00
    0.587399374E+01    0.276542344E+00
Cr    P
    0.228909997E+02   -0.194121061E-01
    0.308550018E+01    0.386188758E+00
    0.121323291E+01    0.676239089E+00
Cr    P
    0.449316807E+00    0.100000000E+01
Cr    P
    0.120675000E+00    0.100000000E+01
Cr    D
    0.437200745E+02    0.220932593E-01
    0.123912427E+02    0.128014388E+00
    0.426394420E+01    0.386529103E+00
    0.155252218E+01    0.641033014E+00
Cr    D
    0.537619295E+00    0.100000000E+01
Cr    D
    0.164931731E+00    0.100000000E+01
Cr    D
    0.510000000E-01    0.100000000E+01
Cr    F
    0.114700000E+01    0.100000000E+01
'''),
'Cr2': gto.basis.parse('''
Cr    S
    0.254477807E+06    0.242904776E-03
    0.381317971E+05    0.188435236E-02
    0.867529306E+04    0.980095844E-02
    0.245500998E+04    0.398250087E-01
    0.799162178E+03    0.129405437E+00
    0.286900215E+03    0.306290018E+00
    0.111254132E+03    0.434628350E+00
    0.438641526E+02    0.224695629E+00
Cr    S
    0.279326692E+03   -0.243633579E-01
    0.862747324E+02   -0.115114954E+00
    0.135557561E+02    0.550922663E+00
    0.569781128E+01    0.536113547E+00
Cr    S
    0.856365826E+01   -0.371230177E+00
    0.139882968E+01    0.116811695E+01
Cr    S
    0.572881711E+00    0.100000000E+01
Cr    S
    0.900961713E-01    0.100000000E+01
Cr    S
    0.341258851E-01    0.100000000E+01
Cr    P
    0.130643989E+04    0.282928007E-02
    0.309253114E+03    0.227766292E-01
    0.989962740E+02    0.105645619E+00
    0.367569165E+02    0.299499449E+00
    0.145666571E+02    0.477062454E+00
    0.587399374E+01    0.276542344E+00
Cr    P
    0.228909997E+02   -0.194121061E-01
    0.308550018E+01    0.386188758E+00
    0.121323291E+01    0.676239089E+00
Cr    P
    0.449316807E+00    0.100000000E+01
Cr    P
    0.120675000E+00    0.100000000E+01
Cr    D
    0.437200745E+02    0.220932593E-01
    0.123912427E+02    0.128014388E+00
    0.426394420E+01    0.386529103E+00
    0.155252218E+01    0.641033014E+00
Cr    D
    0.537619295E+00    0.100000000E+01
Cr    D
    0.164931731E+00    0.100000000E+01
Cr    D
    0.510000000E-01    0.100000000E+01
Cr    F
    0.114700000E+01    0.100000000E+01
''')}

# Remember to check the charge and spin
mol.charge = 0
mol.spin = 0
mol.verbose = 4
mol.build()

mf = scf.UHF(mol)
mf.max_cycle = 1
mf.max_memory = 1000 # MB
mf.kernel()

# read MOs from .fch(k) file
nbf = mf.mo_coeff[0].shape[0]
nif = mf.mo_coeff[0].shape[1]
S = mol.intor_symmetric('int1e_ovlp')
Sdiag = S.diagonal()
alpha_coeff = fch2py('cr2.fch', nbf, nif, Sdiag, 'a')
beta_coeff = fch2py('cr2.fch', nbf, nif, Sdiag, 'b')
mf.mo_coeff = (alpha_coeff, beta_coeff)
# read done

# check if input MOs are orthonormal
check_orthonormal(nbf, nif, mf.mo_coeff[0], S)
check_orthonormal(nbf, nif, mf.mo_coeff[1], S)

dm = mf.make_rdm1()
mf.max_cycle = 10
mf.kernel(dm)

# transform UHF canonical orbitals to UNO
na = np.sum(mf.mo_occ[0]==1)
nb = np.sum(mf.mo_occ[1]==1)
idx, noon, alpha_coeff = uno(nbf, nif, na, nb, mf.mo_coeff[0], mf.mo_coeff[1], S)
alpha_coeff = construct_vir(nbf, nif, idx[1], alpha_coeff, S)
mf.mo_coeff = (alpha_coeff, beta_coeff)
# done transform

# save the UNO into .fch file
os.system('fch_u2r cr2.fch')
os.rename('cr2_r.fch', 'cr2_uno.fch')
py2fch('cr2_uno.fch', nbf, nif, mf.mo_coeff[0], Sdiag, 'a', noon)
# save done
