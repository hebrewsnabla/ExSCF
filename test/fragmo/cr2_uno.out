

******** <class 'pyscf.scf.uhf.UHF'> ********
method = UHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 1
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /home/wsr/ExSCF/test/fragmo/tmp8cvwyuxe
max_memory 1000 MB (current use 87 MB)
number electrons alpha = 24  beta = 24
Set gradient conv threshold to 3.16228e-05
init E= -2085.21751154147
  alpha nocc = 24  HOMO = -0.253876187750209  LUMO = -0.207864800621617
  beta  nocc = 24  HOMO = -0.253204793697658  LUMO = -0.207206079723023
cycle= 1 E= -2085.57268344533  delta_E= -0.355  |g|= 0.533  |ddm|= 1.79
SCF not converged.
SCF energy = -2085.57268344533 after 1 cycles  <S^2> = 0.00010246178  2S+1 = 1.0002049

Orthonormality check: j=  90 i=  19 maxv= 1.42875932E-08

Orthonormality check: j=  90 i=  28 maxv= 1.03534987E-08


******** <class 'pyscf.scf.uhf.UHF'> ********
method = UHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 10
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /home/wsr/ExSCF/test/fragmo/tmp8cvwyuxe
max_memory 1000 MB (current use 166 MB)
number electrons alpha = 24  beta = 24
Set gradient conv threshold to 3.16228e-05
init E= -2085.47596323795
  alpha nocc = 24  HOMO = -0.458954949087602  LUMO = -0.329590679157099
  beta  nocc = 24  HOMO = -0.45806083670035  LUMO = -0.32957534979347
cycle= 1 E= -2085.65846148791  delta_E= -0.182  |g|= 1.74  |ddm|= 3.47
  alpha nocc = 24  HOMO = 0.0444830909710255  LUMO = 0.088893465412938
  beta  nocc = 24  HOMO = 0.0445308067397038  LUMO = 0.0888805718610304
cycle= 2 E= -2084.57134004428  delta_E= 1.09  |g|= 2.28  |ddm|= 9.89
  alpha nocc = 24  HOMO = -0.223180980533691  LUMO = -0.0535927729988084
  beta  nocc = 24  HOMO = -0.222948547967843  LUMO = -0.0537018243885178
cycle= 3 E= -2086.51036051399  delta_E= -1.94  |g|= 0.224  |ddm|= 9.68
  alpha nocc = 24  HOMO = -0.20972141708667  LUMO = 0.0210225756175228
  beta  nocc = 24  HOMO = -0.209605585588624  LUMO = 0.0209898100070853
cycle= 4 E= -2086.56419338043  delta_E= -0.0538  |g|= 0.0999  |ddm|= 0.524
  alpha nocc = 24  HOMO = -0.217542884044072  LUMO = 0.0216271929899856
  beta  nocc = 24  HOMO = -0.217678405036455  LUMO = 0.0216275025086326
cycle= 5 E= -2086.58364547372  delta_E= -0.0195  |g|= 0.0446  |ddm|= 0.299
  alpha nocc = 24  HOMO = -0.218909375473705  LUMO = 0.0243160615408831
  beta  nocc = 24  HOMO = -0.218774868187787  LUMO = 0.0243006512386241
cycle= 6 E= -2086.58975491731  delta_E= -0.00611  |g|= 0.0217  |ddm|= 0.214
  alpha nocc = 24  HOMO = -0.221377927570653  LUMO = 0.0259692144032398
  beta  nocc = 24  HOMO = -0.221509220693357  LUMO = 0.0259877945172051
cycle= 7 E= -2086.59149075775  delta_E= -0.00174  |g|= 0.0116  |ddm|= 0.131
  alpha nocc = 24  HOMO = -0.22579628701298  LUMO = 0.0285400838636278
  beta  nocc = 24  HOMO = -0.225471141858573  LUMO = 0.0285109230607826
cycle= 8 E= -2086.59241098288  delta_E= -0.00092  |g|= 0.00753  |ddm|= 0.134
  alpha nocc = 24  HOMO = -0.227313600341389  LUMO = 0.0290799681459851
  beta  nocc = 24  HOMO = -0.227557587112997  LUMO = 0.0291105803447699
cycle= 9 E= -2086.59259255341  delta_E= -0.000182  |g|= 0.00468  |ddm|= 0.0552
  alpha nocc = 24  HOMO = -0.228653948161573  LUMO = 0.0291759257143589
  beta  nocc = 24  HOMO = -0.22863187845729  LUMO = 0.0291719801348978
cycle= 10 E= -2086.59265126138  delta_E= -5.87e-05  |g|= 0.00117  |ddm|= 0.0375
SCF not converged.
SCF energy = -2086.59265126138 after 10 cycles  <S^2> = 5.2898239  2S+1 = 4.7073661

The orthonormality of Alpha MO after PAO construction:
maxv=    0.0000000000
abs_mean=    0.0000000000
forrtl: severe (24): end-of-file during read, unit -129, file /home/wsr/ExSCF/test/fragmo/cr2.fch
Image              PC                Routine            Line        Source             
fch_u2r            000000000040B85B  Unknown               Unknown  Unknown
fch_u2r            000000000043187E  Unknown               Unknown  Unknown
fch_u2r            0000000000404CCE  MAIN__                     28  fch_u2r.f90
fch_u2r            00000000004037E2  Unknown               Unknown  Unknown
libc-2.32.so       0000146030870152  __libc_start_main     Unknown  Unknown
fch_u2r            00000000004036EE  Unknown               Unknown  Unknown
